pkg/controller/core/leader_aware_reconciler.go

```diff
@@ -0,0 +1,63 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package core
+
+import (
+	"context"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// WithLeadingManager returns a decorating reconcile.Reconciler that discards reconciliation requests
+// for the controllers that are started with the controller.Options.NeedLeaderElection
+// option set to false in non-leading replicas.
+//
+// Starting controllers in non-leading replicas is needed for these that update the data
+// served by the visibility extension API server.
+//
+// This enables to:
+//   - Keep the scheduling decisions under the responsibility of the leading replica alone,
+//     to prevent any concurrency issues.
+//   - Consume requests from the watch event queues, to prevent them from growing indefinitely
+//     in the non-leading replicas.
+//   - Transition to actually reconciling requests in the replica that may acquire
+//     the leader election lease, in case the previously leading replica failed to renew it.
+func WithLeadingManager(mgr ctrl.Manager, reconciler reconcile.Reconciler) reconcile.Reconciler {
+	return &leaderAwareReconciler{
+		elected:  mgr.Elected(),
+		delegate: reconciler,
+	}
+}
+
+type leaderAwareReconciler struct {
+	elected  <-chan struct{}
+	delegate reconcile.Reconciler
+}
+
+var _ reconcile.Reconciler = (*leaderAwareReconciler)(nil)
+
+func (r *leaderAwareReconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	select {
+	case <-r.elected:
+		// The manager has been elected leader, delegate reconciliation to the provided reconciler.
+		return r.delegate.Reconcile(ctx, request)
+	default:
+		// The manager hasn't been elected leader yet, skip reconciliation.
+		return ctrl.Result{}, nil
```

Comment:
If there is a delay in the notification, then would we rely on the Resync to queue the objects for reconciliation again?

Is there a way to trigger Resync when becoming the leader, so we don't have to wait?

Reply:
There might be a delay between the moment the leading replica becomes "inoperative", fails to renew the leader election lease, and the future leading replica acquires the lease, which creates a window during which events might be missed / not taken into account.

It may be possible to trigger a full resync of the cache, but I think it's more suitable for correcting cache drift / entropy, and be a "hammer" solution for preventing event misses. For the latter, I think re-queuing with a constant duration would be sufficient. This would probably mean the queues for non-leading replicas contain the keys for the entire list of existing resources, but I think that's acceptable.

Reply:
Added 4fafe7e63d5b7c8e011762a8fee71a021026ae9c to requeue the request after a constant duration set to the configuration leader election lease duration. PTAL 🙏🏼.

Reply:
> There might be a delay between the moment the leading replica becomes "inoperative", fails to renew the leader election lease, and the future leading replica acquires the lease, which creates a window during which events might be missed / not taken into account.

Can this also happen on the main branch? If this is scenario of missed reconciles happens on main I think we could move it to a follow up, because it seems to add complexity, so maybe fixing the scenario can be decoupled from the visibility server fix?

Reply:
Ok, I think I understand, and it seems this is a new issue (with delayed reconciles due to the change of leader) on this branch. IIUC on main, we would not start the watches on non-leading replicas, so at the moment the replica transitions to leading, the watches would start and we would get the ADDED events, triggering reconciles.

Here, we have the watches active, so we will not get the ADDED events, and we may wait unnecessarily long the the reconciles. Doing this with 15min timeout seems borderline for acceptable, because it is still regression from main. 

Is there a way to trigger reconciles for the Reconcilers when the replica transitions from non-leading to leading?

Reply:
> Ok, I think I understand, and it seems this is a new issue (with delayed reconciles due to the change of leader) on this branch. IIUC on main, we would not start the watches on non-leading replicas, so at the moment the replica transitions to leading, the watches would start and we would get the ADDED events, triggering reconciles.

Right, that's my understanding as well.

> Here, we have the watches active, so we will not get the ADDED events, and we may wait unnecessarily long the the reconciles. Doing this with 15min timeout seems borderline for acceptable, because it is still regression from main.
> 
> Is there a way to trigger reconciles for the Reconcilers when the replica transitions from non-leading to leading?

Mind it's 15 seconds by default, which is the default value for the lease duration that's waited by the non-leading replicas to periodically attempt to acquire leadership. This duration is configurable, by setting `.leaderElection.leaseDuration` in the Kueue controller ConfigMap, if 15 seconds is consider too long.
 
On a side note, re-queuing is documented as the recommended approach to mitigate event misses here:

https://github.com/kubernetes-sigs/controller-runtime/blob/8f8247fdc1c3bc75f2189f28079a9efc4e82d06f/pkg/cache/cache.go#L163-L170

Reply:
ah, 15s seems ok, not sure why I was mistaken by minutes.

Reply:
I thought for a second I had made a mistake on my end, but no 😃.